cdap-app-fabric/src/main/java/io/cdap/cdap/support/task/SupportBundlePipelineInfoTask.java

```diff
@@ -0,0 +1,153 @@
+  private final RemoteProgramRunRecordsFetcher remoteProgramRunRecordsFetcher;
+  private final RemoteProgramLogsFetcher remoteProgramLogsFetcher;
+  private final RemoteMetricsSystemClient remoteMetricsSystemClient;
+  private final List<NamespaceId> namespaces;
+  private final String uuid;
+  private final String app;
+  private final ProgramType programType;
+  private final String programName;
+  private final SupportBundleJob supportBundleJob;
+  private final int maxRunsPerProgram;
+
+  @Inject
+  public SupportBundlePipelineInfoTask(@Assisted String uuid, @Assisted List<NamespaceId> namespaces,
+                                       @Assisted String app, @Assisted File basePath,
+                                       RemoteApplicationDetailFetcher remoteApplicationDetailFetcher,
+                                       RemoteProgramRunRecordsFetcher remoteProgramRunRecordsFetcher,
+                                       RemoteProgramLogsFetcher remoteProgramLogsFetcher,
+                                       @Assisted ProgramType programType, @Assisted String programName,
+                                       RemoteMetricsSystemClient remoteMetricsSystemClient,
+                                       @Assisted SupportBundleJob supportBundleJob, @Assisted int maxRunsPerProgram) {
+    this.uuid = uuid;
+    this.basePath = basePath;
+    this.namespaces = namespaces;
+    this.app = app;
+    this.remoteApplicationDetailFetcher = remoteApplicationDetailFetcher;
+    this.remoteProgramRunRecordsFetcher = remoteProgramRunRecordsFetcher;
+    this.remoteProgramLogsFetcher = remoteProgramLogsFetcher;
+    this.programType = programType;
+    this.programName = programName;
+    this.remoteMetricsSystemClient = remoteMetricsSystemClient;
+    this.supportBundleJob = supportBundleJob;
+    this.maxRunsPerProgram = maxRunsPerProgram;
+  }
+
+  @Override
+  public void collect() throws IOException, NotFoundException {
+    for (NamespaceId namespaceId : namespaces) {
+      List<ApplicationRecord> apps = new ArrayList<>();
+      if (app == null) {
+        apps.addAll(remoteApplicationDetailFetcher.list(namespaceId.getNamespace()).stream()
+                      .map(applicationDetail -> new ApplicationRecord(applicationDetail)).collect(Collectors.toList()));
+      } else {
+        apps.add(new ApplicationRecord(
+          remoteApplicationDetailFetcher.get(new ApplicationId(namespaceId.getNamespace(), app))));
+      }
+
+      for (ApplicationRecord appId : apps) {
+        String application = appId.getName();
+        ApplicationId applicationId = new ApplicationId(namespaceId.getNamespace(), application);
+        File appFolderPath = new File(basePath, appId.getName());
+        DirUtils.mkdirs(appFolderPath);
+        try (FileWriter file = new FileWriter(new File(appFolderPath, appId.getName() + ".json"))) {
+          ApplicationDetail applicationDetail = remoteApplicationDetailFetcher.get(applicationId);
```

Comment:
Handle when it's not found (E.g. user specified app, but not namespace and we are looking through all namespaces for an app)